### PR TITLE
Closes #1792 - `GroupBy` Symbol Table Entry

### LINKED ISSUE #1792: `GroupBySymEntry` Creation
Part of #1740 

Create the `GroupBySymEntry`.
Move the functionality from `GroupBy.__init__` to the server.

As a note, @reuster986 requested that key `GroupBy` properties be directly accessible from the client. The return of the `GroupBySymEntry` creation will include statements to store references to these properties on the client.
- `.permutaion`
- `.segments`
- `.unique_keys`

### PR DESCRIPTION
**Pull request by Ethan-DeBandi99**
Closes #1792 

- Adds a Symbol Table Entry for `GroupBy` objects. These are only created currently in the case when `keys` is provided. If all elements are provided, the object remains client side. Issue #1803 addresses adding support creation from all components.
- `GroupBySymEntry` inherits from `GenSymEntry` in preparation for removing `CompositeSymEntry`. `CompositeSymEntry` currently does not add any information that is not found in `GenSymEntry` and does not appear necessary.
- Updates the client side constructor to send a server message to create the `GroupBy` object. 
- Adds `createGroupBy` message.
- Currently, the `unique_keys` property is not stored on the server. Issue #1804 will address computing Unique Keys on the server. 

This code utilizes components of `Unique`, but does not require the full workflow at this time.

### CODE REVIEW

**Comment by Ethan-DeBandi99** [comment]
Converting to draft until I resolve the bug in gasnet. 

Also, realized I forgot to merge in updates to master and will do that as well.

**Comment by Ethan-DeBandi99** [comment]
Adding a note that #1807 is needed before merging this.

File: arkouda/groupbyclass.py [hunk 1]
```diff
@@ -90,24 +114,25 @@ def unique(
...
+    #     first = True
+    #     for k in pda:
+    #         if first:
+    #             size = k.size
+    #             first = False
+    #         elif k.size != size:
+    #             raise ValueError("Key arrays must all be same size")
+    #         if not hasattr(k, "_get_grouping_keys"):
+    #             raise TypeError(f"{type(k)} does not support grouping")
+    #         grouping_keys.extend(cast(list, k._get_grouping_keys()))
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/groupbyclass.py]: I forgot to remove this. Should be done now.

File: src/MultiTypeSymEntry.chpl [hunk 2]
```diff
@@ -401,6 +402,36 @@ module MultiTypeSymEntry
...
+
+            this.dtype = DType.UNDEF; // not required by groupby, but needs initialization
+            this.itemsize = itemsize; // itemsize will vary and can be accessed from specific keys
+
+            this.ndim = this.segmentsEntry.size; // used as the number of groups
+        }
+
+        override proc getSizeEstimate(): int {
+            return this.segmentsEntry.getSizeEstimate() + this.permEntry.getSizeEstimate();
+        }
```
Comment (pierce314159) [inline src/MultiTypeSymEntry.chpl:432]: I think `getSizeEstimate` should include the other array components as well (`keyNamesEntry`, `keyTypesEntry`, `ukIndEntry`)

**Review by pierce314159 — COMMENTED** [review]
Looks good! The only thing that might need changed is the `getSizeEstimate`

File: arkouda/groupbyclass.py [hunk 3]
```diff
@@ -35,6 +36,29 @@
...
+        # Sequence of groupable arrays
+        nkeys = len(pda)
+        grouping_keys = []
+        first = True
+        for k in pda:
+            if first:
+                size = k.size
+                first = False
+            elif k.size != size:
+                raise ValueError("Key arrays must all be same size")
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/groupbyclass.py:54]: Agreed. This is updated

File: arkouda/groupbyclass.py [hunk 4]
```diff
@@ -236,45 +243,59 @@ def __init__(
...
+                },
             )
-
-        self.length = self.permutation.size
-        self.ngroups = self.segments.size
+            rep_json = json.loads(repmsg)
+            fields = rep_json["groupby"].split()
+            self.name = fields[1]
+            self.length = int(fields[3])
+            self.ngroups = int(fields[4])
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/groupbyclass.py:291]: Yeah, I modeled it after `create_pdarray`. It is also worth noting we may tweak some things moving forward with additional functionality for `GroupBy` as more is moved to the server.

File: src/MultiTypeSymEntry.chpl (hunk 2 above)
Comment (Ethan-DeBandi99) [reply src/MultiTypeSymEntry.chpl:432]: Agreed. I copied this over and forgot to update it. I will add that now

**Review by pierce314159 — APPROVED** [review]

**Pull request merged**